File: libraries/php/tests/WebhooksTest.php
<?php

declare(strict_types=1);

namespace Voidhash\Tests;

use PHPUnit\Framework\TestCase;
use Voidhash\Exception\WebhookVerificationException;
use Voidhash\Webhooks;

final class WebhooksTest extends TestCase
{
    public function testConstructEventVerifiesAndParses(): void
    {
        $payload = '{"hello":"world"}';
        $timestamp = (string) time();
        $signature = 'v1=' . hash_hmac('sha256', $timestamp . '.' . $payload, 'whsec_test');

        $event = Webhooks::constructEvent($payload, [
            'X-Webhook-Event' => ['purchase.completed'],
            'X-Webhook-Timestamp' => [$timestamp],
            'X-Webhook-Signature' => [$signature],
        ], 'whsec_test');

        self::assertSame('purchase.completed', $event->type);
        self::assertSame(['hello' => 'world'], $event->payload);
        self::assertSame((int) $timestamp, $event->timestamp->getTimestamp());
    }

    public function testTamperedSignatureIsRejected(): void
    {
        $payload = '{"hello":"world"}';
        $timestamp = (string) time();
        $signature = substr('v1=' . hash_hmac('sha256', $timestamp . '.' . $payload, 'whsec_test'), 0, -4) . '0000';

        $this->expectException(WebhookVerificationException::class);
        Webhooks::constructEvent($payload, [
            'X-Webhook-Event' => ['purchase.completed'],
            'X-Webhook-Timestamp' => [$timestamp],
            'X-Webhook-Signature' => [$signature],
        ], 'whsec_test');
    }

    public function testStaleTimestampOutsideToleranceIsRejected(): void
    {
        $payload = '{}';
        $timestamp = (string) (time() - 3600);
        $signature = 'v1=' . hash_hmac('sha256', $timestamp . '.' . $payload, 'whsec_test');

        self::assertFalse(Webhooks::verifySignature($payload, $signature, $timestamp, 'whsec_test'));
    }

    public function testMissingHeaderThrows(): void
    {
        $this->expectException(WebhookVerificationException::class);
        Webhooks::constructEvent('{}', [], 'whsec_test');
    }
}

File: libraries/rust/tests/webhooks_test.rs
//! Webhook verification tests, mirroring the vectors from the Node SDK.

use voidhash::webhooks;

fn sign(timestamp: &str, payload: &str, secret: &str) -> String {
    use hmac::{Hmac, Mac};
    use sha2::Sha256;
    let mut mac = Hmac::<Sha256>::new_from_slice(secret.as_bytes()).unwrap();
    mac.update(timestamp.as_bytes());
    mac.update(b".");
    mac.update(payload.as_bytes());
    format!("v1={}", hex::encode(mac.finalize().into_bytes()))
}

#[test]
fn construct_event_verifies_and_parses() {
    let payload = r#"{"hello":"world"}"#;
    let timestamp = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_secs()
        .to_string();
    let signature = sign(&timestamp, payload, "whsec_test");

    let event = webhooks::construct_event(
        payload.as_bytes(),
        "purchase.completed",
        &signature,
        &timestamp,
        "whsec_test",
    )
    .expect("verification succeeds");

    assert_eq!(event.event_type, "purchase.completed");
    assert_eq!(event.payload["hello"], "world");
}

#[test]
fn tampered_signature_is_rejected() {
    let payload = r#"{"hello":"world"}"#;
    let timestamp = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_secs()
        .to_string();
    let signature = {
        let full = sign(&timestamp, payload, "whsec_test");
        format!("{}0000", &full[..full.len() - 4])
    };

    let result = webhooks::construct_event(
        payload.as_bytes(),
        "purchase.completed",
        &signature,
        &timestamp,
        "whsec_test",
    );
    assert!(matches!(
        result,
        Err(webhooks::WebhookVerificationError::InvalidSignature)
    ));
}

#[test]
fn stale_timestamp_outside_tolerance_is_rejected() {
    let stale = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_secs() as i64
        - 3600;
    let timestamp = stale.to_string();
    let signature = sign(&timestamp, "{}", "whsec_test");

    assert!(!webhooks::verify_signature(
        b"{}",
        &signature,
        &timestamp,
        "whsec_test",
        300
    ));
}
